pkg/model-serving-controller/controller/model_serving_controller.go

```diff
@@ -213,6 +213,32 @@ func (c *ModelServingController) deleteModelServing(obj interface{}) {
 		}
 	}
 
+	servingGroupList, err := c.store.GetServingGroupByModelServing(utils.GetNamespaceName(mi))
+	if err != nil && !errors.Is(err, datastore.ErrServingGroupNotFound) {
+		klog.Errorf("cannot get servingGroup of modelServing: %s from map: %v", mi.GetName(), err)
+		return
+	}
+
+	for i := range servingGroupList {
+		services, err := c.getServicesByIndex(GroupNameKey, fmt.Sprintf("%s/%s", mi.Namespace, servingGroupList[i].Name))
+		if err != nil {
+			klog.Errorf("failed to get services for ServingGroup %s: %v", servingGroupList[i].Name, err)
+			return
+		}
+
+		// Because the Service has an OwnerReference set, removing the finalizers allows it to be deleted successfully.
+		for _, svc := range services {
+			svcDeepCopy := svc.DeepCopy()
+			// remove finalizer first to avoid being blocked by deleting headless service
+			utils.RemoveHeadlessServiceProtectionFinalizer(svcDeepCopy)
+			_, err = c.kubeClientSet.CoreV1().Services(mi.Namespace).Update(context.TODO(), svcDeepCopy, metav1.UpdateOptions{})
```

Comment:
I would suggest making this close to the Delete. You can define a delete function, which does update and delete.